temporalnexus/operation.go

```diff
@@ -108,45 +124,159 @@ type SignalWorkflowInput struct {
+	// If not set, SignalWorkflow will be used.
+	// Workflow function to map this operation to. The operation input maps directly to workflow input.
+	// The workflow name is resolved as it would when using this function in client.ExecuteOperation.
+	// GetOptions must be provided when setting this option. Mutually exclusive with Handler.
+	Workflow func(workflow.Context, I) (O, error)
+	// Options for starting the workflow. Must be set if Workflow is set. Mutually exclusive with Handler.
+	// If the options returned include a workflow ID, it must match the workflow ID set in the signal input. That
+	// workflow ID must be deterministically generated from the input in order for the operation to be idempotent as
+	// the request to start the operation may be retried. TaskQueue is optional and defaults to the current worker's
+	// task queue.
+	GetStartOptions func(context.Context, I, nexus.StartOperationOptions) (client.StartWorkflowOptions, error)
+
+	// Handler for starting a workflow with a different input than the operation. Mutually exclusive with
+	// GetSignalInput, Workflow and GetStartOptions.
+	Handler func(context.Context, I, nexus.StartOperationOptions) (nexus.HandlerStartOperationResult[O], error)
+}
+
+type workflowSignalOperation[I, O any] struct {
+	nexus.UnimplementedOperation[I, O]
+
+	options WorkflowSignalOperationOptions[I, O]
+}
+
 // NewWorkflowSignalOperation is a helper for creating a synchronous nexus.Operation to deliver a signal, linking the
 // signal to a Nexus operation. Request ID from the Nexus options is propagated to the workflow to ensure idempotency.
+// The operation is complete as soon as the signal is delivered and returns no value.
 //
 // NOTE: Experimental
-func NewWorkflowSignalOperation[T any](
+func NewWorkflowSignalOperation[I any](
 	name string,
-	getSignalInput func(context.Context, T, nexus.StartOperationOptions) SignalWorkflowInput,
-) nexus.Operation[T, nexus.NoValue] {
-	return NewSyncOperation(name, func(ctx context.Context, c client.Client, in T, options nexus.StartOperationOptions) (nexus.NoValue, error) {
-		signalInput := getSignalInput(ctx, in, options)
-
-		if options.RequestID != "" {
-			ctx = context.WithValue(ctx, internal.NexusOperationRequestIDKey, options.RequestID)
-		}
+	getSignalInput func(context.Context, I, nexus.StartOperationOptions) (WorkflowSignalInput, error),
+) nexus.Operation[I, nexus.NoValue] {
+	return &workflowSignalOperation[I, nexus.NoValue]{
+		options: WorkflowSignalOperationOptions[I, nexus.NoValue]{
+			Name:           name,
+			GetSignalInput: getSignalInput,
+		},
+	}
+}
 
-		links, err := convertNexusLinks(options.Links, GetLogger(ctx))
-		if err != nil {
-			return nil, err
-		}
-		ctx = context.WithValue(ctx, internal.NexusOperationLinksKey, links)
+// NewWorkflowSignalWithStartOperation is a helper for creating a synchronous nexus.Operation to deliver a signal to a
```

Comment:
I think having `NewWorkflowSignalWithStartOperation` be synchronous is not very intuitive for users, I would expect a signalWithStart operation to be an async operation that attaches to the existing workflow

Reply:
To me, signal-with-start is about the signal, not about the workflow run. Users will typically use this API to ensure that there's a workflow running to receive the signal.